daemon/build.go

```diff
@@ -104,13 +105,19 @@ func (daemon *Daemon) pullForBuilder(ctx context.Context, name string, authConfi
 // Every call to GetImageAndReleasableLayer MUST call releasableLayer.Release() to prevent
 // leaking of layers.
 func (daemon *Daemon) GetImageAndReleasableLayer(ctx context.Context, refOrID string, opts backend.GetImageAndLayerOptions) (builder.Image, builder.ReleaseableLayer, error) {
-	if !opts.ForcePull {
-		image, _ := daemon.GetImage(refOrID)
+	id, _ := daemon.GetImageID(refOrID)
+	refIsID := id.String() == refOrID // detect if ref is an ID to skip pulling
```

Comment:
I don't know what you mean by it being incorrect. It uses `GetImageID` specifically to only have single code-path for this detection.

This returns the previous(checked 17.03) `docker pull && docker run` semantics as it is a bugfix to `17.06` multi-stage build regression. If you want to change the semantics we can discuss that(though I think I disagree) but shouldn't be part of this PR.